File: pkg/nuclide-python-rpc/python/outline.py
# Copyright (c) 2015-present, Facebook, Inc.
# All rights reserved.
#
# This source code is licensed under the license found in the LICENSE file in
# the root directory of this source tree.

import jedi
from parso.python.tree import ImportFrom


def serialize_names(names, visited):
    return list(filter(None, [serialize_name(n, visited) for n in names]))


def serialize_name(name, visited):
    if name in visited:
        # Some class structures seem to cause cycles :(
        return None
    visited.add(name)

    result = {}

    if name.type == 'class' or name.type == 'function':
        result['name'] = name.name
        try:
            result['params'] = [p.name for p in name.params]
        except Exception:
            # ".params" appears to be quite flaky.
            # e.g: https://github.com/davidhalter/jedi/issues/1031
            result['params'] = []
        result['children'] = (
            serialize_names(name.defined_names(), visited)
            if name.type == 'class' else [])
    elif name.type == 'statement':
        # Don't include underscore assignments in outline.
        if name.name == '_':
            return None
        result['name'] = name.name
    else:
        return None

    try:
        definition = name._name.tree_name.get_definition()
    except AttributeError:
        return None

    # Do not include imported definitions.
    if isinstance(definition, ImportFrom):
        return None

    (start_line, start_column) = definition.start_pos
    (end_line, end_column) = definition.end_pos

    result['start'] = {
        'line': start_line,
        'column': start_column,
    }

    result['end'] = {
        'line': end_line,
        'column': end_column,
    }

    result['kind'] = name.type

    return result


def is_toplevel(node):
    try:
        # Jedi bug: parent() throws for certain nodes.
        return node.parent().type == 'module'
    except AttributeError:
        return False


def get_outline(src, contents):
    names = jedi.api.names(source=contents, path=src, all_scopes=True)
    # Only iterate through top-level definitions.
    names = filter(is_toplevel, names)
    return serialize_names(names, visited=set())

File: pkg/nuclide-python-rpc/lib/outline.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.itemsToOutline = itemsToOutline;

function _simpleTextBuffer() {
  const data = require("simple-text-buffer");

  _simpleTextBuffer = function () {
    return data;
  };

  return data;
}

function _tokenizedText() {
  const data = require("../../../modules/nuclide-commons/tokenized-text");

  _tokenizedText = function () {
    return data;
  };

  return data;
}

/**
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the LICENSE file in
 * the root directory of this source tree.
 *
 *  strict-local
 * @format
 */
function itemToOutlineTree(mode, item) {
  switch (item.kind) {
    case 'class':
      return classToOutlineTree('all', item);

    case 'function':
      return functionToOutlineTree(item);

    case 'statement':
      return statementToOutlineTree(mode, item);
  }
}

function classToOutlineTree(mode, item) {
  return Object.assign({
    tokenizedText: [(0, _tokenizedText().keyword)('class'), (0, _tokenizedText().whitespace)(' '), (0, _tokenizedText().method)(item.name)],
    representativeName: item.name,
    children: itemsToOutline(mode, item.children)
  }, itemToPositions(item));
}

function functionToOutlineTree(item) {
  return Object.assign({
    tokenizedText: [(0, _tokenizedText().keyword)('def'), (0, _tokenizedText().whitespace)(' '), (0, _tokenizedText().method)(item.name), (0, _tokenizedText().plain)('('), ...argsToText(item.params || []), (0, _tokenizedText().plain)(')')],
    representativeName: item.name,
    children: []
  }, itemToPositions(item));
}

function statementToOutlineTree(mode, item) {
  if (mode === 'none') {
    return null;
  }

  const name = item.name; // Only show initialization of constants, which according to python
  // style are all upper case.

  if (mode === 'constants' && name !== name.toUpperCase()) {
    return null;
  }

  return Object.assign({
    tokenizedText: [(0, _tokenizedText().plain)(name)],
    representativeName: name,
    children: []
  }, itemToPositions(item));
}

function argsToText(args) {
  const result = [];

  function startArg() {
    if (result.length > 0) {
      result.push((0, _tokenizedText().plain)(','));
      result.push((0, _tokenizedText().whitespace)(' '));
    }
  }

  args.forEach(arg => {
    startArg();

    if (arg.startsWith('**')) {
      result.push((0, _tokenizedText().plain)('**'));
      result.push((0, _tokenizedText().param)(arg.slice(2)));
    } else if (arg.startsWith('*')) {
      result.push((0, _tokenizedText().plain)('*'));
      result.push((0, _tokenizedText().param)(arg.slice(1)));
    } else {
      result.push((0, _tokenizedText().param)(arg));
    }
  });
  return result;
}

function itemToPositions(item) {
  const {
    start,
    end
  } = item;
  return {
    startPosition: new (_simpleTextBuffer().Point)(start.line - 1, start.column),
    // Outline's endPosition is inclusive, while Jedi's is exclusive.
    // By decrementing the end column, we avoid situations where
    // two items are highlighted at once. End column may end up as -1,
    // which still has the intended effect.
    endPosition: new (_simpleTextBuffer().Point)(end.line - 1, end.column - 1)
  };
}

function itemsToOutline(mode, items) {
  if (!items || items.length === 0) {
    return [];
  }

  const result = [];
  items.map(i => itemToOutlineTree(mode, i)).forEach(tree => {
    if (tree) {
      result.push(tree);
    }
  });
  return result;
}